dxp/resource_injection.md

```diff
@@ -0,0 +1,106 @@
+
+You can see `<liferay-util:dynamic-include>`'s TLD [here](https://github.com/liferay/liferay-portal/blob/5908bcbc5a5f731f13456ed3e8441ab3f2910123/util-taglib/src/META-INF/liferay-util.tld#L42).
+
+# `TopHeadResources`
+
+[`TopHeadResources`](https://github.com/liferay/liferay-portal/blob/cc9be6b4dbf289902cb8edb89f1b38b1334e6284/modules/apps/frontend-js/frontend-js-top-head-extender-api/src/main/java/com/liferay/frontend/js/top/head/extender/TopHeadResources.java) is a frontend infrastructure service built on top of [`DynamicInclude`](https://github.com/liferay/liferay-portal/blob/973e6199844436c22d237e2daf4821d0f9af5362/portal-kernel/src/com/liferay/portal/kernel/servlet/taglib/DynamicInclude.java).
+
+It is used to inject JavaScript resources in the `<head>` tag of the rendered HTML so that is is available to the page.
+
+It distinguishes between guest and authenticated resources (the latter including the former). This allows serving a limited subset of JavaScript files for guest users that, usually, don't need the full set of JavaScript files to use the portal.
+
+As multiple `TopHeadResources` can be deployed, they are prioritized according to their [`service.ranking`](https://docs.osgi.org/javadoc/r2/org/osgi/framework/Constants.html#SERVICE_RANKING).
+
+## How to define `TopHeadResources`
+
+You can define [`TopHeadResources`](https://github.com/liferay/liferay-portal/blob/cc9be6b4dbf289902cb8edb89f1b38b1334e6284/modules/apps/frontend-js/frontend-js-top-head-extender-api/src/main/java/com/liferay/frontend/js/top/head/extender/TopHeadResources.java) using Java code like in [AUITopHeadResources](https://github.com/liferay/liferay-portal/blob/73dc0b6400829ed66ff6f17807814ac6381308e3/modules/apps/frontend-js/frontend-js-aui-web/src/main/java/com/liferay/frontend/js/aui/web/internal/servlet/AUITopHeadResources.java#L93), but you can also use an easier [declarative syntax](https://github.com/liferay/liferay-portal/blob/e38fb4fe9b1bf87af0de4280c468da0821866882/modules/apps/frontend-js/frontend-js-web/bnd.bnd#L4).
+
+### Declarative Syntax
+
+The declarative syntax is used in `bnd.bnd` files and makes use of three different headers:
+
+-   `Liferay-JS-Resources-Top-Head`: this specifies a comma delimited list of JavaScript resources to include always. The path of the files is relative to `src/main/resource/META-INF/resources` as they are fetched through the bundle's servlet context.
+-   `Liferay-JS-Resources-Top-Head-Authenticated`: this specifies a comma delimited list of JavaScript resources to include **only when the user has been authenticated**. The path of the files is relative to `src/main/resources/META-INF/resources` as they are fetched through the bundle's servlet context.
+-   `Liferay-Top-Head-Weight`: the value of this property is used as [`service.ranking`](https://docs.osgi.org/javadoc/r2/org/osgi/framework/Constants.html#SERVICE_RANKING) to prioritize the associated `TopHeadResources`.
```

Comment:
Ah cool. I wanted to know how to do this the other day but didn't know this property existed.